components/script/animations.rs

```diff
@@ -171,103 +204,195 @@ impl Animations {
         set.clear_canceled_animations();
     }
 
-    fn handle_new_animations(set: &mut ElementAnimationSet, update: &mut AnimationsUpdate) {
+    fn handle_new_animations(
+        &self,
+        set: &mut ElementAnimationSet,
+        now: f64,
+        pipeline_id: PipelineId,
+    ) {
         for animation in set.animations.iter_mut() {
             animation.is_new = false;
         }
 
         for transition in set.transitions.iter_mut() {
             if transition.is_new {
-                // TODO(mrobinson): We need to properly compute the elapsed_time here
-                // according to https://drafts.csswg.org/css-transitions/#event-transitionevent
-                update.add_event(
-                    transition.node,
-                    transition.property_animation.property_id().name().into(),
+                self.add_transition_event(
+                    transition,
                     TransitionOrAnimationEventType::TransitionRun,
-                    0.,
+                    now,
+                    pipeline_id,
                 );
                 transition.is_new = false;
             }
         }
     }
-}
 
-impl MallocSizeOf for Animations {
-    fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
-        self.sets.read().size_of(ops) + self.have_running_animations.size_of(ops)
+    /// Ensure that all nodes with new animations are rooted. This should be called
+    /// immediately after a restyle, to ensure that these addresses are still valid.
+    #[allow(unsafe_code)]
+    fn root_newly_animating_dom_nodes(
+        &self,
+        sets: &FxHashMap<OpaqueNode, ElementAnimationSet>,
+        js_runtime: *mut JSRuntime,
```

Comment:
Can we accept a `&Runtime`argument instead so we don't pass around raw pointers? We can use [rt](https://doc.servo.org/mozjs/rust/struct.Runtime.html#method.rt) to get the right pointer from it.

Reply:
:+1: